Extension/src/LanguageServer/cppBuildTaskProvider.ts

```diff
@@ -0,0 +1,346 @@
+    private writeEmitter  = new vscode.EventEmitter<string>();
+    private closeEmitter = new vscode.EventEmitter<number>();
+    public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+    public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+
+    constructor(private command: string, private args: string[], private options: cp.ExecOptions | undefined) {
+    }
+
+
+    open(_initialDimensions: vscode.TerminalDimensions | undefined): void {
+        telemetry.logLanguageServerEvent("cppBuildTaskStarted");
+        // At this point we can start using the terminal.
+        this.writeEmitter.fire("Starting build...\r\n");
+        this.doBuild();
+    }
+
+    close(): void {
+        // The terminal has been closed. Shutdown the build.
+    }
+
+    private async doBuild(): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            // Do build.
+            let activeCommand: string = util.resolveVariables(this.command, this.AdditionalEnvironment);
+            this.args.forEach(value => {
+                let temp: string = util.resolveVariables(value, this.AdditionalEnvironment);
+                if (temp && temp.includes(" ")) {
+                    temp = "\"" + temp + "\"";
+                }
+                activeCommand = activeCommand + " " + temp;
+            });
+            if (this.options?.cwd) {
+                this.options.cwd = util.resolveVariables(this.options.cwd, this.AdditionalEnvironment);
+            }
+            cp.exec(activeCommand, this.options, (_error, stdout, _stderr) => {
+                if (_error) {
+                    telemetry.logLanguageServerEvent("cppBuildTaskError");
+                    this.writeEmitter.fire("Build finished with error:\r\n");
+                    this.writeEmitter.fire(stdout.toString());
+                    reject();
+                } else {
+                    this.writeEmitter.fire(stdout.toString());
+                    this.writeEmitter.fire("\r\nBuild finished successfully.\r\n");
+                    resolve(0);
+                }
+            });
+        }).finally (() => {
+            this.closeEmitter.fire(0);
+        });
+    }
+
+    private get AdditionalEnvironment(): { [key: string]: string | string[] } | undefined {
+
+        const editor: vscode.TextEditor | undefined = vscode.window.activeTextEditor;
+        if (!editor) {
+            return undefined;
+        }
+        const fileDir: vscode.WorkspaceFolder | undefined = vscode.workspace.getWorkspaceFolder(editor.document.uri);
+        if (!fileDir) {
+            return undefined;
+        }
+        const file: string = editor.document.fileName;
+        return {
+            "file": file,
+            "fileDirname": fileDir.uri.fsPath,
+            "fileBasenameNoExtension": path.parse(file).name,
+            "workspaceFolder": fileDir.uri.fsPath};
+    }
+}
+
+export function getRawTasksJson(): Promise<any> {
```

Comment:
I know you didn't write this function, but...  There is a `checkFileExists` function in common.ts that is await-able.  If that was used instead, this function could be made await-able and simplified.

Reply:
`getRawTasksJson` is modified now.

Reply:
I was hoping we could remove creation of a new promise here, as it's not necessary if this function is async, and calls await on checkFileExists (instead of then()).  This function could be verysimple:

export async function getRawTasksJson(): Promise<any> {
    const path: string | undefined = getTasksJsonPath();
    if (!path) {
        return {};
    }
    const fileExists: boolean = await util.checkFileExists(path);
    if (!fileExists) {
        return {};
    }

    const fileContents: string = await util.readFileText(path);
    let rawTasks: any = {};
    try {
        rawTasks = jsonc.parse(fileContents);
    } catch (error) {
        throw new Error(failedToParseTasksJson);
    }
    return rawTasks;
}